### PR TITLE
feat: add cli artifacts compare command

### PR DESCRIPTION
**Pull request by mdonnalley**
### What does this PR do?

Adds `cli artifacts compare` command for comparing `command-snapshot.json` and `schemas/` for each plugin (including JIT) in the CLI. Throws error if breaking change is found in any plugin

Example output

```
~/repos/salesforcecli/cli (sh/add-jit-plugins) » ~/repos/salesforcecli/plugin-release-management/bin/dev cli:artifacts:compare
Warning: No schemas found for salesforcecli/plugin-telemetry@2.1.0
Warning: No schemas found for salesforcecli/plugin-data@2.1.24
Warning: No schemas found for salesforcecli/plugin-signups@1.4.0
Warning: No schemas found for salesforcecli/plugin-trust@2.4.0
Warning: No schemas found for salesforcecli/plugin-info@2.4.0
Warning: No schemas found for salesforcecli/plugin-telemetry@2.0.7
Warning: No schemas found for salesforcecli/plugin-data@2.1.23
Warning: No schemas found for salesforcecli/plugin-trust@2.3.2
Warning: No schemas found for salesforcecli/plugin-info@2.3.6
Warning: No schemas found for salesforcecli/plugin-signups@1.3.2
Warning: No schemas found for salesforcecli/plugin-functions@1.17.4
Warning: No schemas found for salesforcecli/plugin-functions@1.17.3
=== @salesforce/plugin-settings

Current: 1.2.0
Previous: 1.1.15

Snapshot Changes
alias:list No Changes
alias:set No Changes
alias:unset No Changes
config:get No Changes
config:list No Changes
config:set No Changes
config:unset No Changes

Schema Changes

alias:unset
  - ❌ definitions.alias-1501530572-0-211-1501530572-0-1533943985149 was not found in current schema
  - definitions.SfError.properties.cause.properties.cause.$ref was changed from #/definitions/alias-1501530572-0-211-1501530572-0-1533943985149 to #/definitions/alias-1283098157-0-211-1283098157-0-1533943985149
  - definitions.alias-1283098157-0-211-1283098157-0-1533943985149 was added to current schema

alias:set
  - ❌ definitions.alias-1501530572-0-211-1501530572-0-1533943985149 was not found in current schema
  - definitions.SfError.properties.cause.properties.cause.$ref was changed from #/definitions/alias-1501530572-0-211-1501530572-0-1533943985149 to #/definitions/alias-1283098157-0-211-1283098157-0-1533943985149
  - definitions.alias-1283098157-0-211-1283098157-0-1533943985149 was added to current schema

=== @salesforce/plugin-deploy-retrieve

Current: 1.8.0
Previous: 1.7.2

Snapshot Changes
deploy No Changes
deploy:metadata No Changes
deploy:metadata:cancel No Changes
deploy:metadata:preview No Changes
deploy:metadata:quick No Changes
deploy:metadata:report No Changes
deploy:metadata:resume No Changes
deploy:metadata:validate No Changes
retrieve:metadata No Changes
retrieve:metadata:preview No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-env

Current: 2.0.1
Previous: 1.6.0

Snapshot Changes
❌ Removed Commands: [
  'env:create:sandbox',
  'env:create:scratch',
  'env:delete:sandbox',
  'env:delete:scratch',
  'env:resume:sandbox',
  'env:resume:scratch'
]
env:display No Changes
env:list No Changes
env:open No Changes

Schema Changes

env:resume:scratch
  - ❌ env:resume:scratch was not found in current schema

env:resume:sandbox
  - ❌ env:resume:sandbox was not found in current schema

env:delete:scratch
  - ❌ env:delete:scratch was not found in current schema

env:delete:sandbox
  - ❌ env:delete:sandbox was not found in current schema

env:create:scratch
  - ❌ env:create:scratch was not found in current schema

env:create:sandbox
  - ❌ env:create:sandbox was not found in current schema

=== @salesforce/plugin-functions

Current: 1.17.4
Previous: 1.17.3

Snapshot Changes
deploy:functions No Changes
env:compute:collaborator:add No Changes
env:create:compute No Changes
env:delete No Changes
env:log No Changes
env:log:tail No Changes
env:logdrain:add No Changes
env:logdrain:list No Changes
env:logdrain:remove No Changes
env:var:get No Changes
env:var:list No Changes
env:var:set No Changes
env:var:unset No Changes
generate:function No Changes
login:functions No Changes
login:functions:jwt No Changes
logout:functions No Changes
run:function No Changes
run:function:start No Changes
run:function:start:container No Changes
run:function:start:local No Changes
whoami:functions No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-generate

Current: 1.1.0
Previous: 1.0.45

Snapshot Changes
generate:project No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-info

Current: 2.4.0
Previous: 2.3.6

Snapshot Changes
doctor No Changes
info:releasenotes:display No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-limits

Current: 2.3.0
Previous: 2.2.4

Snapshot Changes
limits:api:display No Changes
limits:recordcounts:display No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-schema

Current: 2.2.3
Previous: 2.2.2

Snapshot Changes
sobject:describe No Changes
sobject:list No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-login

Current: 1.1.17
Previous: 1.1.16

Snapshot Changes
login No Changes
login:org No Changes
login:org:jwt No Changes
logout No Changes
logout:org No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-org

Current: 2.3.0
Previous: 2.2.25

Snapshot Changes
force:org:clone No Changes
force:org:create No Changes
force:org:delete No Changes
force:org:status No Changes
org:create:sandbox No Changes
org:create:scratch No Changes
org:delete:sandbox No Changes
org:delete:scratch No Changes
org:display No Changes
org:list No Changes
org:open No Changes
org:resume:sandbox No Changes
org:resume:scratch No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-sobject

Current: 0.1.0
Previous: 0.0.42

Snapshot Changes
schema:generate:field No Changes
schema:generate:platformevent No Changes
schema:generate:sobject No Changes
schema:generate:tab No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-telemetry

Current: 2.1.0
Previous: 2.0.7

Snapshot Changes
telemetry No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-trust

Current: 2.4.0
Previous: 2.3.2

Snapshot Changes
plugins:trust:verify No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-data

Current: 2.1.24
Previous: 2.1.23

Snapshot Changes
data:create:record No Changes
data:delete:bulk No Changes
data:delete:record No Changes
data:export:tree No Changes
data:get:record No Changes
data:import:tree No Changes
data:query No Changes
data:query:resume No Changes
data:resume No Changes
data:update:record No Changes
data:upsert:bulk No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-user

Current: 2.3.0
Previous: 2.2.1

Snapshot Changes
force:user:password:generate No Changes
force:user:permset:assign No Changes
force:user:permsetlicense:assign No Changes
org:assign:permset No Changes
org:assign:permsetlicense No Changes
org:create:user No Changes
org:display:user No Changes
org:generate:password No Changes
org:list:users No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-custom-metadata

Current: 2.1.1
Previous: 2.1.0

Snapshot Changes
cmdt:generate:field No Changes
cmdt:generate:fromorg No Changes
cmdt:generate:object No Changes
cmdt:generate:record No Changes
cmdt:generate:records No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-community

Current: 2.2.1
Previous: 2.2.0

Snapshot Changes
community:create No Changes
community:list:template No Changes
community:publish No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-packaging

Current: 1.14.1
Previous: 1.14.0

Snapshot Changes
package1:version:create No Changes
package1:version:create:get No Changes
package1:version:display No Changes
package1:version:list No Changes
package:convert No Changes
package:create No Changes
package:delete No Changes
package:install No Changes
package:install:report No Changes
package:installed:list No Changes
package:list No Changes
package:uninstall No Changes
package:uninstall:report No Changes
package:update No Changes
package:version:create No Changes
package:version:create:list No Changes
package:version:create:report No Changes
package:version:delete No Changes
package:version:displayancestry No Changes
package:version:list No Changes
package:version:promote No Changes
package:version:report No Changes
package:version:update No Changes

Schema Changes
No changes have been detected.

=== @salesforce/plugin-signups

Current: 1.4.0
Previous: 1.3.2

Snapshot Changes
org:create:shape No Changes
org:create:snapshot No Changes
org:delete:shape No Changes
org:delete:snapshot No Changes
org:get:snapshot No Changes
org:list:shape No Changes
org:list:snapshot No Changes

Schema Changes
No changes have been detected.

Error: Breaking changes found in artifacts
```

### What issues does this PR fix or reference?
@W-12489891@

### CODE REVIEW

File: messages/cli.artifacts.compare.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "description": "Look for breaking changes in artifacts (schemas and snapshots) from plugins",
```
Comment (jshackell-sfdc) [inline messages/cli.artifacts.compare.json:2]:
```suggestion
  "description": "Look for breaking changes in artifacts (schemas and snapshots) from plugins.",
```

**Review by jshackell-sfdc — CHANGES_REQUESTED** [review]
@mdonnalley : teeny edit, thx.

File: src/commands/cli/artifacts/compare.ts [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+    }
+  }
+
+  private async getCurrentPlugins(): Promise<Record<string, string>> {
+    return this.flags.current ? this.getPluginsForVersion(this.current) : this.filterPlugins(this.packageJson);
+  }
+
+  private async getPluginsForVersion(version: string): Promise<Record<string, string>> {
+    if (!this.versions.includes(version)) {
+      throw messages.createError('error.VersionNotFound', [version]);
```
Comment (iowillhoit) [inline src/commands/cli/artifacts/compare.ts:469]: Is this needed here since we have already checked that the version exists in `resolveVersions()`?

**Comment by iowillhoit** [comment]
## QA NOTES

> I've still got a few things to test, here is a WIP QA

- 🟡 `description` text is repeated twice in `--help` output.
- 🟡 I initially tried running this command from the `bin/dev` of `plugin-release-management` which obviously failed. Might be worth checking that the current directory is a "cli", maybe by checking `oclif.bin`?
- 🟢 Flag additions log (not warn)
- 🟢 New commands log (not warn)
- 🟢 Able to compare against old `sf` with `--previous`
- 🟢 Collects correct plugin versions from a `--previous` cli
- 🟢 Detects removed commands in snapshot
- 🟢 Detects removed commands in schema

File: src/commands/cli/artifacts/compare.ts (hunk 2 above)
Comment (mdonnalley) [reply src/commands/cli/artifacts/compare.ts:469]: Oops - just a vestige of a pervious implementation. I'll remove it.

**Comment by iowillhoit** [comment]
## QA NOTES

> I've still got a few things to test, here is a WIP QA

- 🟡 `description` text is repeated twice in `--help` output.
- 🟡 I initially tried running this command from the `bin/dev` of `plugin-release-management` which obviously failed. Might be worth checking that the current directory is a "cli", maybe by checking `oclif.bin`?
- 🟢 Flag additions log (not warn)
- 🟢 New commands log (not warn)
- 🟢 Able to compare against old `sf` with `--previous`
- 🟢 Collects correct plugin versions from a `--previous` cli
- 🟢 Detects removed commands in snapshot
- 🟢 Detects removed commands in schema

# QA UPDATE

- 🟢 Error is thrown if `current` is newer than `previous`
- 🟢 Able to compare against old `sf` with `--current`
- 🟢 Collects correct plugin versions from a `--current` cli
- 🟡 Would it be worth moving `Warning: No schemas found for ...` to the Summary at the bottom?
- 🟢 Throws an error if `current` or `previous` does not exist
- 🟢 Works with `jitPlugins` 🎉 Tested with `"@salesforce/plugin-packaging": "1.14.4"`
- 🟡 The `semver.lt` check could potentially fail if someone was manually adding a prerelease version and didn't bump major/minor/patch also. Not sure if this is worth adding logic for or just something to keep in mind. Here is an example where the `current` (beta) is technically less than the `previous`:

```sh
Running: bin/dev cli:artifacts:compare
Current Version: 1.65.0-beta.0 # Manually changed in package.json
Previous Version: 1.65.0
Error: Current version 1.65.0-beta.0 must be newer than previous version 1.65.0.
```

**Review by iowillhoit — APPROVED** [review]

**Pull request merged**